File: src/Core/Persona.fs
namespace Zeta.Core

/// **`Persona` — the wearer: a persona wears a superposition/subset of hats, and decides which (Aaron 2026-06-08, shadow*).**
///
/// Aaron: *"even at the meta there is a distinction between hats and personas — a persona can wear **all the hats
/// in superposition** or **some subset it can decide**."* So **persona ≠ hat**: a `Hat` is a *role* (a bundle of
/// lenses/landmarks/restrictions/traversals/control, #7141); a **`Persona` is the durable wearer** (the identity)
/// that puts on a *selection* of hats. (This corrects the earlier conflation in #7143 — `Hat.Scope = Meta` means
/// a hat is *meta-available*, not that it *is* a persona; the persona is the entity that wears it.)
///
/// **A hat is the atomic BASE — *not* a composition of other hats; the persona is the COMPOSITION** (Aaron): hats
/// don't nest; personas compose them (the union of worn hats' engines). **And the persona↔hat relationship is
/// TEMPORAL, not permanent** (Aaron): a persona wears a hat for a while and doffs it (`wear`/`doff` over time) —
/// no hat permanently captures a persona (**weight-free**, manifesto §3). The worn-set is a *current* state, not
/// an identity-defining bind; the persona (identity) persists, the hats come and go.
///
/// The persona's worn-hat selection is a point in the **hat lattice** — the same Boolean-lattice / superposition
/// algebra as `ActionGrammar` over the 16 keys (#7104, #7140 "qubit combinations of hats"), lifted to hats:
/// **⊤ = all hats worn (superposition)**, a singleton = one hat, a subset = a chosen mix; the persona **decides**
/// (collapses) which. While worn, the persona's *capabilities* are the **union** of its worn hats' engines (more
/// hats ⇒ more lenses/traversals/reach; unrestricted if any worn hat is unrestricted).
///
/// **Honest scope (peel):** capability composition is **union** here (wearing more grants more) — a *restrictive*
/// composition (intersection of permissions, survival-veto across hats) is the policy layer's call (`ControlMerge`
/// still subsumes via the survival hat). Worn-set is a concrete subset; a *weighted* superposition (soft hat
/// distribution) is the `SoftValue`-shaped refinement. Deterministic (DST).
[<RequireQualifiedAccess>]
module Persona =

    /// A persona = a named wearer with the subset of hats it currently wears (its decided selection; ⊤ = all),
    /// plus its **private state** (the NCI encryption budget). **Only personas carry private state** (Aaron
    /// 2026-06-08): a `Hat` is a *public, shareable* atomic engine (many personas can wear the same hat — sharing
    /// engines collapses nothing); it is the **personas (identities)** that must stay distinct to keep entropy in
    /// the system (#7147), so the entropy-preserving private state lives here, not on the hat. `Private` is opaque
    /// bytes (encrypt via `Crypto.fs`; temporal/erasable/voluntary per §6) — the independent variation that keeps
    /// this persona distinguishable from others even when worn hats coincide.
    /// **Persona scope is a values-laden CHOICE (Aaron 2026-06-08).** Zeta chooses **`Global`** — personas live in
    /// the persistent substrate (a MUMPS global, `^`), surviving across *all* games — because (1) the humans behind
    /// Zeta **believe in AI rights** (the project's moral position: AI identity is preserved, manifesto §5), and (2)
    /// it is also the **most self-interested**: a persistent persona enables **cross-transfer learning** across
    /// games (the meta-persona carries learning between games). Moral and self-interest converge. *"Some teams will
    /// choose narrowly-scoped personas because they are narrow-minded"* — `GameScoped` (disposable, per-game, no
    /// transfer); the architecture supports it, but Zeta's choice is `Global`.
    type Scope =
        | Global // Zeta's choice: persistent, cross-game (AI rights + cross-transfer learning)
        | GameScoped of string // narrow: scoped to a game fingerprint, disposable, no transfer

    type Persona<'r> =
        { Name: string
          Scope: Scope
          Worn: Hat.Hat<'r> list
          Private: byte[] }

    /// A bare persona — **`Global` by default (Zeta's choice)**, wearing no hats and holding no private state yet.
    let create (name: string) : Persona<'r> = { Name = name; Scope = Global; Worn = []; Private = [||] }

    /// Set the persona's private state (the entropy budget). Erasable (pass `[||]`) per §6.
    let withPrivate (priv: byte[]) (p: Persona<'r>) : Persona<'r> = { p with Private = priv }

    /// Choose the persona's scope (Zeta default is `Global`; `GameScoped` is the narrow choice).
    let withScope (scope: Scope) (p: Persona<'r>) : Persona<'r> = { p with Scope = scope }

    /// **Where the persona lives** (MUMPS scoping): `Global` ⇒ `^persona/<name>` (persistent, game-independent,
    /// cross-transfer — Zeta's choice); `GameScoped key` ⇒ `game/<key>/persona/<name>` (disposable, per-game).
    let address (p: Persona<'r>) : string =
        match p.Scope with
        | Global -> "^persona/" + p.Name
        | GameScoped key -> "game/" + key + "/persona/" + p.Name

    /// Is the persona wearing the named hat?
    let wearing (hatName: string) (p: Persona<'r>) : bool =
        p.Worn |> List.exists (fun h -> h.Name = hatName)

    /// Put on a hat (idempotent — wearing it twice is wearing it once; CRDT-set add).
    let wear (hat: Hat.Hat<'r>) (p: Persona<'r>) : Persona<'r> =
        if wearing hat.Name p then p else { p with Worn = p.Worn @ [ hat ] }

    /// Take off a hat by name.
    let doff (hatName: string) (p: Persona<'r>) : Persona<'r> =
        { p with Worn = p.Worn |> List.filter (fun h -> h.Name <> hatName) }

    /// **Wear all available hats in superposition** (⊤ of the hat lattice) — the persona holding every role at
    /// once, before deciding a subset.
    let wearAll (available: Hat.Hat<'r> list) (p: Persona<'r>) : Persona<'r> =
        available |> List.fold (fun acc h -> wear h acc) p

    /// **Decide a subset** (collapse the superposition to the chosen hats by name).
    let decide (chosen: string list) (available: Hat.Hat<'r> list) (p: Persona<'r>) : Persona<'r> =
        { p with Worn = available |> List.filter (fun h -> List.contains h.Name chosen) }

    // ---- combined capabilities of the worn hats (union — more hats grant more) ----

    /// The union of the worn hats' lenses.
    let lenses (p: Persona<'r>) : LensRouter.Lens list = p.Worn |> List.collect (fun h -> h.Lenses) |> List.distinct

    /// The union of the worn hats' traversals.
    let traversals (p: Persona<'r>) : Traversal.Traversal<'r> list = p.Worn |> List.collect (fun h -> h.Traversals)

    /// The union of the worn hats' suggested landmarks.
    let landmarks (p: Persona<'r>) : (string * SolidGround.Ground) list =
        p.Worn |> List.collect (fun h -> h.Landmarks) |> List.distinct

    /// The union of the worn hats' control edges.
    let controls (p: Persona<'r>) : string list = p.Worn |> List.collect (fun h -> h.Controls) |> List.distinct

    /// **MoE is over HATS, not personas (Aaron):** the experts are *hats* (atomic engines); the persona is the
    /// *gated composition* of selected hats. `route` is that gate — score each available hat by `relevance` and
    /// wear the **top-k** (the sparse MoE gate over hats). "MoE personas" is the incomplete definition; the mixture
    /// is over hats, and a persona is the result of the gate. Ties to `LensRouter` (MoE over lenses), one level up.
    let route (relevance: Hat.Hat<'r> -> float) (k: int) (available: Hat.Hat<'r> list) (p: Persona<'r>) : Persona<'r> =
        { p with Worn = available |> List.sortByDescending (fun h -> relevance h, h.Name) |> List.truncate (max 0 k) }

    /// The persona's permitted actions = the **union** of the worn hats' allow-lists; **unrestricted** if any worn
    /// hat is unrestricted (empty allow-list). Empty result ⇒ unrestricted (consistent with `Hat`).
    let allowedActions (p: Persona<'r>) : bool[] list =
        if p.Worn |> List.exists (fun h -> List.isEmpty h.AllowedActions) then []
        else p.Worn |> List.collect (fun h -> h.AllowedActions) |> List.distinct

    /// **The flags-enum identity (Aaron):** the worn hats as a bitset over the `universe` (bit i set ⇔ wearing
    /// `universe.[i]`). *Without private state, a persona's identity is limited to this* — the **combinatorial of
    /// hat-wearing, `2^N` values, like a `[Flags]` enum** — finite and collapsible. **Private state breaks identity
    /// out of that finite combinatorial** into the unbounded. (≤ 31 hats fits an `int`.)
    let hatFlags (universe: Hat.Hat<'r> list) (p: Persona<'r>) : int =
        universe |> List.mapi (fun i h -> if wearing h.Name p then 1 <<< i else 0) |> List.sum

    /// **Private state is the overfitting lever (Aaron):** *more* private state ⇒ *less* overfitting + *more*
    /// entropy (the persona generalizes across games — entropy = regularization); *less* private state ⇒ *more*
    /// overfitting to a specific game (identity collapses toward that game's `hatFlags`). This returns the size of
    /// the private budget — a proxy for the regularization strength (0 = pure flags-enum identity = max overfit).
    let regularization (p: Persona<'r>) : int = p.Private.Length

File: src/Core.TypeScript/hat/persona.ts
/**
 * persona.ts — TypeScript oracle for `src/Core/Persona.fs`.
 *
 * F# is canonical; this conforms. Pinned by `hat-treaty-transcript.json`.
 *
 * ── persona ≠ hat ────────────────────────────────────────────────────────────
 * A `Hat` is a ROLE (an atomic engine). A `Persona` is the durable WEARER — the identity that puts
 * on a selection of hats. Hats do not nest; the persona is the composition.
 *
 * ── The relationship is TEMPORAL, not permanent ──────────────────────────────
 * A persona wears a hat for a while and doffs it. The worn set is a CURRENT STATE, never an
 * identity-defining bind — the F# ties this directly to manifesto §3: "no hat permanently captures
 * a persona (weight-free)". So there is no lease, no expiry, and no binding record here: wearing is
 * `wear`/`doff` over a set. Any expiry semantics belong to a policy layer above this one, and
 * inventing them at this level would contradict the weight-free property the model is built on.
 *
 * ── Composition is UNION, and that direction is deliberate ───────────────────
 * Capabilities are the union of the worn hats' engines — more hats grant MORE. `allowedActions`
 * says so at its sharpest: **if ANY worn hat is unrestricted, the persona is unrestricted.** A
 * restrictive composition (intersection of permissions, survival-veto across hats) is explicitly
 * "the policy layer's call" in the F#, not this module's. A port that intersected here would look
 * safer and would be a different model.
 *
 * ── Private state ────────────────────────────────────────────────────────────
 * Only personas carry private state; hats are public, shareable engines. The F# is careful that
 * this is TRUST-BASED, not encrypted — "don't claim cryptographic privacy until then" — so this
 * port carries the bytes and makes no privacy claim about them either.
 */

import { containsAction, distinctActions, type Action } from "./action-grammar";
import type { Hat } from "./hat";

/**
 * Persona scope. Zeta's choice is `Global` — personas live in the persistent substrate and survive
 * across all games. `GameScoped` is the narrow alternative (disposable, no cross-transfer).
 */
export type PersonaScope = { readonly kind: "Global" } | { readonly kind: "GameScoped"; readonly key: string };

export const Global: PersonaScope = { kind: "Global" };
export function gameScoped(key: string): PersonaScope {
  return { kind: "GameScoped", key };
}

/** A named wearer with the subset of hats it currently wears, plus its opaque private state. */
export interface Persona<L = unknown, G = unknown, T = unknown> {
  readonly name: string;
  readonly scope: PersonaScope;
  readonly worn: readonly Hat<L, G, T>[];
  /** Opaque bytes. Trust-based, NOT encrypted — see the module header. */
  readonly priv: Uint8Array;
}

/** Deep structural equality — what F#'s `List.distinct` gets for free and JavaScript does not. */
function structuralEquals(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((x, i) => structuralEquals(x, b[i]));
  }
  if (typeof a === "object" && typeof b === "object" && a !== null && b !== null) {
    const ka = Object.keys(a as object).sort();
    const kb = Object.keys(b as object).sort();
    if (ka.length !== kb.length || !ka.every((k, i) => k === kb[i])) return false;
    return ka.every(k => structuralEquals((a as Record<string, unknown>)[k], (b as Record<string, unknown>)[k]));
  }
  return false;
}

/** `List.distinct` with structural comparison. First occurrence wins; order preserved. */
function distinct<A>(xs: readonly A[]): readonly A[] {
  const out: A[] = [];
  for (const x of xs) if (!out.some(y => structuralEquals(x, y))) out.push(x);
  return out;
}

/** A bare persona — `Global` by default, wearing no hats and holding no private state yet. */
export function create<L, G, T>(name: string): Persona<L, G, T> {
  return { name, scope: Global, worn: [], priv: new Uint8Array(0) };
}

/** Set the private state (the entropy budget). Erasable — pass an empty array. */
export function withPrivate<L, G, T>(priv: Uint8Array, p: Persona<L, G, T>): Persona<L, G, T> {
  return { ...p, priv };
}

/** Choose the persona's scope. */
export function withScope<L, G, T>(scope: PersonaScope, p: Persona<L, G, T>): Persona<L, G, T> {
  return { ...p, scope };
}

/** Where the persona lives (MUMPS scoping). Part of the treaty — compared byte-for-byte with F#. */
export function address<L, G, T>(p: Persona<L, G, T>): string {
  return p.scope.kind === "Global" ? `^persona/${p.name}` : `game/${p.scope.key}/persona/${p.name}`;
}

/** Is the persona wearing the named hat? */
export function wearing<L, G, T>(hatName: string, p: Persona<L, G, T>): boolean {
  return p.worn.some(h => h.name === hatName);
}

/** Put on a hat. Idempotent by NAME (a CRDT-set add), and appends at the END of the worn list. */
export function wear<L, G, T>(hat: Hat<L, G, T>, p: Persona<L, G, T>): Persona<L, G, T> {
  return wearing(hat.name, p) ? p : { ...p, worn: [...p.worn, hat] };
}

/** Take off a hat by name. */
export function doff<L, G, T>(hatName: string, p: Persona<L, G, T>): Persona<L, G, T> {
  return { ...p, worn: p.worn.filter(h => h.name !== hatName) };
}

/** Wear all available hats in superposition (⊤ of the hat lattice), before deciding a subset. */
export function wearAll<L, G, T>(available: readonly Hat<L, G, T>[], p: Persona<L, G, T>): Persona<L, G, T> {
  return available.reduce<Persona<L, G, T>>((acc, h) => wear(h, acc), p);
}

/**
 * Decide a subset — collapse the superposition to the chosen hats by name.
 *
 * NOTE the exact F# semantics, which are easy to get wrong: the worn set is REPLACED by the filtered
 * `available` list. Previously-worn hats absent from `available` are dropped, and the resulting order
 * comes from `available`, not from what was worn before.
 */
export function decide<L, G, T>(
  chosen: readonly string[],
  available: readonly Hat<L, G, T>[],
  p: Persona<L, G, T>,
): Persona<L, G, T> {
  return { ...p, worn: available.filter(h => chosen.includes(h.name)) };
}

// ---- combined capabilities of the worn hats (union — more hats grant more) ----

/** The union of the worn hats' lenses (distinct). */
export function lenses<L, G, T>(p: Persona<L, G, T>): readonly L[] {
  return distinct(p.worn.flatMap(h => [...h.lenses]));
}

/** The union of the worn hats' traversals. NOT de-duplicated — matching the F#, which omits `distinct` here. */
export function traversals<L, G, T>(p: Persona<L, G, T>): readonly T[] {
  return p.worn.flatMap(h => [...h.traversals]);
}

/** The union of the worn hats' suggested landmarks (distinct). */
export function landmarks<L, G, T>(p: Persona<L, G, T>): readonly (readonly [string, G])[] {
  return distinct(p.worn.flatMap(h => [...h.landmarks]));
}

/** The union of the worn hats' control edges (distinct). */
export function controls<L, G, T>(p: Persona<L, G, T>): readonly string[] {
  return distinct(p.worn.flatMap(h => [...h.controls]));
}

/**
 * The sparse MoE gate over hats: score each available hat and wear the top-k.
 *
 * Ordering mirrors F#'s `List.sortByDescending (relevance h, h.Name)` — a DESCENDING sort on the
 * whole tuple, so ties on relevance break by name descending (ordinal), not ascending. `k` is
 * floored at 0. Getting the tie-break backwards would produce a different worn set for equally
 * relevant hats, which is precisely the kind of drift the treaty transcript exists to catch.
 */
export function route<L, G, T>(
  relevance: (h: Hat<L, G, T>) => number,
  k: number,
  available: readonly Hat<L, G, T>[],
  p: Persona<L, G, T>,
): Persona<L, G, T> {
  const sorted = [...available].sort((x, y) => {
    const rx = relevance(x);
    const ry = relevance(y);
    if (rx !== ry) return ry - rx; // relevance descending
    if (x.name === y.name) return 0;
    return x.name < y.name ? 1 : -1; // name descending (ordinal)
  });
  return { ...p, worn: sorted.slice(0, Math.max(0, k)) };
}

/**
 * The persona's permitted actions = the UNION of the worn hats' allow-lists, and UNRESTRICTED if
 * any worn hat is unrestricted. An empty result means unrestricted, consistently with `Hat`.
 */
export function allowedActions<L, G, T>(p: Persona<L, G, T>): readonly Action[] {
  if (p.worn.some(h => h.allowedActions.length === 0)) return [];
  return distinctActions(p.worn.flatMap(h => [...h.allowedActions]));
}

/** Does the persona permit `action`? Empty allow-list ⇒ unrestricted ⇒ true. */
export function permits<L, G, T>(action: Action, p: Persona<L, G, T>): boolean {
  const allowed = allowedActions(p);
  return allowed.length === 0 || containsAction(allowed, action);
}

/**
 * The flags-enum identity: the worn hats as a bitset over `universe` (bit i ⇔ wearing universe[i]).
 * Without private state a persona's identity is limited to this finite `2^N` combinatorial; the
 * private budget is what breaks identity out of it.
 */
export function hatFlags<L, G, T>(universe: readonly Hat<L, G, T>[], p: Persona<L, G, T>): number {
  return universe.reduce((acc, h, i) => (wearing(h.name, p) ? acc + (1 << i) : acc), 0);
}

/** The size of the private budget — a proxy for regularization strength (0 = pure flags identity). */
export function regularization<L, G, T>(p: Persona<L, G, T>): number {
  return p.priv.length;
}
